### PR TITLE
Adds first version of auth logic

### PR DESCRIPTION
**Pull request by Boychenko**

### CODE REVIEW

File: src/Likvido.ApiAuth.Client/HttpRequestHeadersExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            return requestHeaders;
+        }
+
+        public static HttpRequestHeaders SetUserInfoHeader(
+            this HttpRequestHeaders requestHeaders,
+            IEnumerable<Claim> claims,
+            string headerName = AuthConstants.UserInfoHeader)
+        {
+            requestHeaders.Remove(headerName);
+            requestHeaders.Add(headerName, UserInfoHeaderSerializer.Serialize( claims));
```
Comment (spewu) [inline src/Likvido.ApiAuth.Client/HttpRequestHeadersExtensions.cs]:
```suggestion
            requestHeaders.Add(headerName, UserInfoHeaderSerializer.Serialize(claims));
```

File: src/Likvido.ApiAuth.Client/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# Likvido.ApiAuth.Client
+Library to simplify sending user info header from a client side
+## Usage
+### WebClient
+```
+using Likvido.ApiAuth.Client;
```
Comment (spewu) [inline src/Likvido.ApiAuth.Client/README.md:7]:
```suggestion
using System.Security.Claims;
using Likvido.ApiAuth.Client;
```

File: src/Likvido.ApiAuth.Client/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+var claims = new List<Claim>
+{
+    new Claim(ClaimTypes.Role, "Admin"),
+    new Claim(ClaimTypes.Role, "Creditor"),
+    new Claim(ClaimTypes.NameIdentifier, userId),
+    new Claim("whatever", "1"),
+    new Claim("namespace/whatever", "2"),
+};
+
+//optional if aren't happy with default mapping or want to add some speicifcmaaping
```
Comment (spewu) [inline src/Likvido.ApiAuth.Client/README.md]:
```suggestion
//optional if you aren't happy with default mapping or you want to add some specific mapping
```

File: src/Likvido.ApiAuth.Client/README.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+//if no mapping exists original claim will be used in token
```
Comment (spewu) [inline src/Likvido.ApiAuth.Client/README.md]:
```suggestion
//if no mapping exists, then the original claim will be used in the token
```

File: src/Likvido.ApiAuth.Client/README.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+var mapping = TokenClaims.Default;
+mapping.AddTokenClaimToClaim("role", ClaimTypes.Role);
+
+var client = new WebClient();
+//The overload method without "userId" is this case "ClaimTypes.NameIdentifier" will be used to detect a user id
```
Comment (spewu) [inline src/Likvido.ApiAuth.Client/README.md]:
```suggestion
//Using the overload method without the "userId" will result in using the "ClaimTypes.NameIdentifier" to detect the user id
```

File: src/Likvido.ApiAuth.Client/README.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+client.Headers.SetUserInfoHeader(userId, claims, mapping);
+```
+
+### HttpClient
+```
+using Likvido.ApiAuth.Client;
+//same as WebClient apart from a couple of lines
```
Comment (spewu) [inline src/Likvido.ApiAuth.Client/README.md]:
```suggestion
//the same as with the WebClient except for a couple of lines
```

File: src/Likvido.ApiAuth.Client/README.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+var client = new HttpClient();
+client.DefaultRequestHeaders.SetUserInfoHeader(userId, claims, mapping);
+```
+### HttpRequestMessage
+```
+using Likvido.ApiAuth.Client;
+//same as WebClient apart from a couple of lines
```
Comment (spewu) [inline src/Likvido.ApiAuth.Client/README.md]:
```suggestion
//the same as with the WebClient except for a couple of lines
```

File: src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+            if (!hasId)
+            {
+                claims.Add(new Claim(ClaimTypes.NameIdentifier, jsonUserId));
+            }
+
+            return new ClaimsIdentity(claims, authType);
+        }
+
+        public static string  Serialize(string userId, IEnumerable<Claim> claims, TokenClaims tokenClaims = null)
```
Comment (spewu) [inline src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs]:
```suggestion
        public static string Serialize(string userId, IEnumerable<Claim> claims, TokenClaims tokenClaims = null)
```

File: src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            MaxDepth = 3
+        };
+
+        public static ClaimsIdentity Deserialize(string value, TokenClaims tokenClaims = null, string authType = AuthConstants.UserInfoHeader)
+        {
+            tokenClaims = tokenClaims ?? TokenClaims.Default;
+            var rawString = Encoding.UTF8.GetString(Convert.FromBase64String(value));
+            var doc = JsonDocument.Parse(rawString, DocumentOptions);
+            var claims = new List<Claim>();
+            var jsonUserId = doc.RootElement.GetProperty("userId").GetString();
```
Comment (spewu) [inline src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs]: Maybe we should handle the case where no `userId` has been added? - throw a specific exception?

File: src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+                else
+                {
+                    userInfo.Claims.Add(claimName, claimGroup.First().Value);
+                }
+            }
+
+            return Convert.ToBase64String(Encoding.UTF8.GetBytes(JsonSerializer.Serialize(userInfo, SerializerOptions)));
+        }
+
+        public static string  Serialize(IEnumerable<Claim> claims, TokenClaims tokenClaims = null)
```
Comment (spewu) [inline src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs]:
```suggestion
        public static string Serialize(IEnumerable<Claim> claims, TokenClaims tokenClaims = null)
```

File: src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+                {
+                    userInfo.Claims.Add(claimName, claimGroup.Select(c => c.Value).ToArray());
+                }
+                else
+                {
+                    userInfo.Claims.Add(claimName, claimGroup.First().Value);
+                }
+            }
+
+            return Convert.ToBase64String(Encoding.UTF8.GetBytes(JsonSerializer.Serialize(userInfo, SerializerOptions)));
```
Comment (spewu) [inline src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs]: Maybe extract this into a private re-usable method?

File: src/Likvido.ApiAuth.Server/ApiKeyAuthMiddleware.cs [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+            else
+            {
+                _logger.ApiKeyAuthorizationFailureOccurred(
+                    "Api Key not provided.",
+                    new Dictionary<string, object> { ["HeaderName"] = options.HeaderName });
+                await WriteUnauthorized(context).ConfigureAwait(false);
+            }
+        }
+
+        private static Task WriteUnauthorized(HttpContext context)
```
Comment (spewu) [inline src/Likvido.ApiAuth.Server/ApiKeyAuthMiddleware.cs]: What is the reason for returning a task here, and calling it as async?

File: src/Likvido.ApiAuth.Server/README.md [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+Default header name `X-API-Key`.
+
+**SHOULD go before UseAuthentication or UserInfoMiddlewareExtensions**
+
+```
+    public class ApiKeyAuthOptions : ApiKeyAuthOptionsBase
+    {
+        public override string GetUserName(string apiKey)
+        {
+            //Add you key - consuming app mapping logic
```
Comment (spewu) [inline src/Likvido.ApiAuth.Server/README.md]:
```suggestion
            //Add your key - consuming app mapping logic
```

File: src/Likvido.ApiAuth.Server/README.md [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        options.DefaultScheme = AuthConstants.UserInfoHeader;
+    }).AddScheme<UserInfoHeaderAuthenticationSchemeOptions, UserInfoHeaderAuthenticationHandler>(
+        AuthConstants.UserInfoHeader,
+        op => { });
+}
+
+// This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
+public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+{
+    //ommited for briefity
```
Comment (spewu) [inline src/Likvido.ApiAuth.Server/README.md]:
```suggestion
    //omitted for brevity
```

File: src/Likvido.ApiAuth.Server/README.md [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    app.UseRouting();
+
+    app.UseApiKeyAuthMiddleware();
+
+    app.UseAuthentication();
+    app.UseAuthorization();
+
+    //ommited for briefity
```
Comment (spewu) [inline src/Likvido.ApiAuth.Server/README.md]:
```suggestion
    //omitted for brevity
```

**Review by spewu — CHANGES_REQUESTED** [review]

File: src/Likvido.ApiAuth.Server/ApiKeyAuthMiddleware.cs (hunk 12 above)
Comment (Boychenko) [reply src/Likvido.ApiAuth.Server/ApiKeyAuthMiddleware.cs]: I've been writing error details to a response initially. Then I decided to stop doing this. A consumer can use `StatusCodePages` middleware to return properly shaped error responses. 
But you are right no need in return Task now

File: src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs (hunk 11 above)
Comment (Boychenko) [reply src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs]: I forgot to reuse the whole `Serialize`. Now fixed

File: src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs (hunk 9 above)
Comment (Boychenko) [reply src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs]: right made deserialization stricter with better errors

**Review by spewu — APPROVED** [review]
Sweet!

**Pull request merged**